core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -113,47 +115,44 @@ class TransactionStateManager(brokerId: Int,
+  def getAndMaybeAddTransactionState(transactionalId: String,
+                                     txnMetadata: TransactionMetadata = null): Either[Errors, Option[CoordinatorEpochAndTxnMetadata]]
+  = inReadLock(stateLock) {
     val partitionId = partitionFor(transactionalId)
 
-    // we only need to check leaving partition set but not loading partition set since there are three possible cases:
-    //    1) it is not in the loading partitions set, hence safe to return NONE
-    //    2) it is in the loading partitions with a smaller epoch, hence safe to return NONE
-    //    3) it is in the loading partition with a larger epoch, return NONE is also fine as it
-    //       indicates the metadata is not exist at least for now.
-    //
-    //    4) it is NOT possible to be in the loading partition with the same epoch
-    if (leavingPartitions.exists(_.txnPartitionId == partitionId))
-      return None
+    if (loadingPartitions.exists(_.txnPartitionId == partitionId))
+      return Left(Errors.COORDINATOR_LOAD_IN_PROGRESS)
 
-    transactionMetadataCache.get(partitionId).flatMap { cacheEntry =>
-      cacheEntry.metadataPerTransactionalId.get(transactionalId) match {
-        case null => None
-        case txnMetadata => Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata))
-      }
-    }
-  }
-
-  /**
-   * Add a new transaction metadata, or retrieve the metadata if it already exists with the associated transactional id
-   * along with the current coordinator epoch for that belonging transaction topic partition
-   */
-  def addTransaction(transactionalId: String, txnMetadata: TransactionMetadata): CoordinatorEpochAndTxnMetadata = {
-    val partitionId = partitionFor(transactionalId)
+    if (leavingPartitions.exists(_.txnPartitionId == partitionId))
+      Right(None)
 
     transactionMetadataCache.get(partitionId) match {
-      case Some(txnMetadataCacheEntry) =>
-        val currentTxnMetadata = txnMetadataCacheEntry.metadataPerTransactionalId.putIfNotExists(transactionalId, txnMetadata)
-        if (currentTxnMetadata != null) {
-          CoordinatorEpochAndTxnMetadata(txnMetadataCacheEntry.coordinatorEpoch, currentTxnMetadata)
-        } else {
-          CoordinatorEpochAndTxnMetadata(txnMetadataCacheEntry.coordinatorEpoch, txnMetadata)
+      case Some(cacheEntry) =>
+        cacheEntry.metadataPerTransactionalId.get(transactionalId) match {
+          case null =>
+            if (txnMetadata != null) {
+              val currentTxnMetadata = cacheEntry.metadataPerTransactionalId.putIfNotExists(transactionalId, txnMetadata)
+              if (currentTxnMetadata != null) {
+                Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, currentTxnMetadata)))
+              } else {
+                Right(Some(CoordinatorEpochAndTxnMetadata(cacheEntry.coordinatorEpoch, txnMetadata)))
+              }
+            } else {
+              Right(None)
```

Comment:
I don't think so: `None` is used to indicate there is no txn metadata and the caller is not trying to create a new one, different callers will handle `None` differently, e.g. return `NOT_COORDINATOR`, return `INVALID_PID`, throw an exception.